app/src/main/java/io/homeassistant/companion/android/widgets/ButtonWidgetConfigureActivity.kt

```diff
@@ -38,72 +48,116 @@ class ButtonWidgetConfigureActivity : Activity() {
         val intent = Intent()
         intent.action = ButtonWidget.RECEIVE_DATA
         intent.component = ComponentName(context, ButtonWidget::class.java)
+
+        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
+
+        // Analyze and send service and domain
+        val serviceText = context.widget_text_config_service.text.toString()
+        val domain = services[serviceText]?.domain ?: serviceText.split(".", limit = 2)[0]
+        val service = services[serviceText]?.service ?: serviceText.split(".", limit = 2)[1]
         intent.putExtra(
             ButtonWidget.EXTRA_DOMAIN,
-            services.getItem(context.service.selectedItemPosition)?.domain
+            domain
         )
         intent.putExtra(
             ButtonWidget.EXTRA_SERVICE,
-            services.getItem(context.service.selectedItemPosition)?.service
-        )
-        intent.putExtra(
-            ButtonWidget.EXTRA_ENTITY_ID,
-            entities.getItem(entity_id.selectedItemPosition)?.entityId
+            service
         )
+
+        // Fetch and send label and icon
         intent.putExtra(
             ButtonWidget.EXTRA_LABEL,
-            label.text.toString()
+            context.label.text.toString()
         )
         intent.putExtra(
             ButtonWidget.EXTRA_ICON,
             context.widget_config_spinner.selectedItemId.toInt()
         )
-        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
+
+        // Analyze and send service data
+        val serviceDataMap = HashMap<String, Any>()
+        for (i in 0 until dynamicFields.size) {
+            val fieldText = dynamicFieldAdapter.getText(i)
+
+            // Don't store data that's empty (or just whitespace)
+            if (!fieldText.isBlank()) {
+                serviceDataMap[dynamicFields[i].second] = fieldText
+            }
+        }
+
+        intent.putExtra(
+            ButtonWidget.EXTRA_SERVICE_DATA,
+            Gson().toJson(serviceDataMap)
+        )
+
         context.sendBroadcast(intent)
 
         // Make sure we pass back the original appWidgetId
         setResult(RESULT_OK, Intent().putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId))
         finish()
     }
 
+    private val dropDownOnFocus = View.OnFocusChangeListener { view, hasFocus ->
+        if (hasFocus && view is AutoCompleteTextView) {
+            view.showDropDown()
+        }
+    }
+
+    private val serviceTextWatcher: TextWatcher = (object : TextWatcher {
+        override fun afterTextChanged(p0: Editable?) {
+            val serviceText: String = p0.toString()
+
+            if (services.keys.contains(serviceText)) {
+                Log.d(TAG, "Valid domain and service--processing dynamic fields")
+
+                // Make sure there are not already any dynamic fields created
+                // This can happen if selecting the drop-down twice or pasting
+                widget_config_fields_layout.removeAllViews()
```

Comment:
The `notifyDataSetChanged()` should be enough to ensure the ui is up to date.

Reply:
Yes, this should have been removed. Sorry for the sloppiness.

Reply:
This was actually changed to clear out the Array of fields that the RecyclerView references.
Changed in commit 29fc233